federation/pkg/federation-controller/service/dns.go

```diff
@@ -16,25 +16,244 @@ limitations under the License.
+	// Quinton: Pseudocode....
+	// See https://github.com/kubernetes/kubernetes/pull/25107#issuecomment-218026648
+	// For each service we need the following DNS names:
+	// mysvc.myns.myfed.svc.z1.r1.mydomain.com  (for zone z1 in region r1)
+	//         - an A record to IP address of specific shard in that zone (if that shard exists and has healthy endpoints)
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.r1.mydomain.com  (if a healthy shard does not exist in zone z1)
+	// mysvc.myns.myfed.svc.r1.federation
+	//         - a set of A records to IP addresses of all healthy shards in region r1, if one or more of these exist
+	//         - OR a CNAME record to the next level up, i.e. mysvc.myns.myfed.svc.mydomain.com (if no healthy shards exist in region r1)
+	// mysvc.myns.myfed.svc.federation
+	//         - a set of A records to IP addresses of all healthy shards in all regions, if one or more of these exist.
+	//         - no record (NXRECORD response) if no healthy shards exist in any regions)
+	//
+	// For each cached service, cachedService.lastState tracks the current known state of the service, while cachedService.appliedState contains
+	// the state of the service when we last successfully sync'd it's DNS records.
+	// So this time around we only need to patch that (add new records, remove deleted records, and update changed records.
+	//
+	if s == nil {
+		return fmt.Errorf("nil ServiceController passed to ServiceController.ensureDnsRecords(clusterName: %s, cachedService: %v)", clusterName, cachedService)
+	}
+	if cachedService == nil {
+		return fmt.Errorf("nil cachedService passed to ServiceController.ensureDnsRecords(clusterName: %s, cachedService: %v)", clusterName, cachedService)
+	}
+	serviceName := cachedService.lastState.Name
+	namespaceName := cachedService.lastState.Namespace
+	if zoneNames, regionName, err := s.getClusterZoneNames(clusterName); err != nil {
+		return err
+	} else {
+		if dnsZoneName, err := s.getFederationDNSZoneName(); err != nil {
+			return err
+		} else {
+			zoneEndpoints, regionEndpoints, globalEndpoints, err := s.getHealthyEndpoints(clusterName, cachedService)
+			if err != nil {
+				return err
+			}
+
+			commonPrefix := serviceName + "." + namespaceName + "." + s.federationName + ".svc"
+			// dnsNames is the path up the DNS search tree, starting at the leaf
+			dnsNames := []string{
+				commonPrefix + "." + zoneNames[0] + "." + regionName + "." + dnsZoneName, // zone level - TODO might need other zone names for multi-zone clusters
+				commonPrefix + "." + regionName + "." + dnsZoneName,                      // region level, one up from zone level
+				commonPrefix + "." + dnsZoneName,                                         // global level, one up from region level
+				"", // nowhere to go up from global level
+			}
+
+			endpoints := [][]string{zoneEndpoints, regionEndpoints, globalEndpoints}
```

Comment:
In that case the watch on the kubernetes service status triggers this function to be called, and we should remove the DNS record (modulo the bug in line 180 above, which I'm busy addressing now).  At least that's what's supposed to happen, but I think you might have uncovered another bug here. I need a special case for zero healthy endpoints.  Stand by for a fix...